### PR TITLE
Style guide

### PR DESCRIPTION
**Pull request by Andrewwango**
Motivation, included in this PR:

> The DeepInverse community maintains a high, opionionated standard of code and docs in order to provide a didactic library that leads the field of imaging, rather than a collection of code files.
> All contributors are responsible for using their human judgement to uphold this standard, which is especially important in the era of LLM coding.
> The purpose of this style guide is to help any devs (experienced, new, maintainers, LLMs, agents) always stick with best practice while contributing **and** reviewing.

It will help with many things

- cleaner, faster PR review by humans
- cleaner LLM contributions without writing the same prompt every time
- more helpful LLM co-review

The added style guide was co-created by the maintainers. It will evolve!

Also clean up docstring guidelines, better contributing page structure

### Checks to be done before submitting your PR

- [x] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` and `ruff check .` run successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### LLM policy

LLM usage is ok, but not PRs generated 100% by AI. See our [LLM policy](https://deepinv.org/contributing.html#llm-policy) Tick below as appropriate:

- [x] I did not use LLM tools to write the code
- [ ] I, a human, wrote code, and LLM tools helped me to write part of the code.
- [ ] An LLM tool wrote all of the code.
- [ ] An agent submitted the PR and wrote the description.

### CODE REVIEW

File: docs/source/contributing.rst [hunk 1]
```diff
@@ -226,30 +230,56 @@ DeepInverse allows contributions where code is partially written by an LLM. Howe
 
 Therefore, DeepInverse does not welcome PRs a) consisting fully of LLM-generated code, or b) that are submitted by an AI agent, or an agent acting on behalf of a human, especially for first time contributors. DeepInverse maintainers may close a PR if they suspect that the PR is AI-generated. Therefore, to help maintainers trust that you are a human coder, we request that, when submitting a PR, you tick whether an LLM or AI agent helped you write the code, or generated it fully, and/or submitted the PR.
 
+.. _style_guides:
+
+Contributing style guides
+-------------------------
+
+The DeepInverse community maintains a high, opionionated standard of code and docs in order to provide a didactic library that leads the field of imaging, rather than a collection of code files.
```
Comment (nucli-vicky) [inline docs/source/contributing.rst]: `opionionated` -> `opinionated`

File: docs/source/contributing.rst [hunk 2]
```diff
@@ -226,30 +230,56 @@ DeepInverse allows contributions where code is partially written by an LLM. Howe
...
+
+- Favour concise docs, and prioritise readability over completeness. Write in concise technical English, not prose. For example:
+
+      Acquisition angles in degrees. Returns ``None`` for vector-based geometries, for which the acquisition trajectory is fully described by `self.projection_geometry["Vectors"]`.
+
+  can be much more concisely written as:
+
+      Astra projection geometry angles tensor in degrees. If Astra vector geom, return None.
+
+- Properly describe each parameter and return, along with a type annotations for each `:param` field, as shown below:
```
Comment (nucli-vicky) [inline docs/source/contributing.rst:267]: should we explicitly mention to keep inline comments concise and only where necessary? 

Mainly targeted at LLM-use which often makes a lot (and long) inline comments

File: docs/source/contributing.rst [hunk 3]
```diff
@@ -294,8 +321,65 @@ Below is a minimal working example of a typical docstring that includes all thes
         def __init__(self, in_channels: int, out_channels: int, pretrained: bool = None):
             pass
 
+**Guidelines are similar for examples.** A good example should teach the user why and how they should use a particular functionality.
+
+- If you introduce a reconstruction method, detail the algorithm, cite it and link to the original code;
```
Comment (nucli-vicky) [inline docs/source/contributing.rst:328]: linking to the original code is something a lot of examples still need to do, right?

File: docs/source/contributing.rst [hunk 4]
```diff
@@ -294,8 +321,65 @@ Below is a minimal working example of a typical docstring that includes all thes
...
+  Therefore, adding an optimisation method requires to check that it converges to the limit point; adding a neural network from an external library should check metric on a dataset, etc. Limit cases should be added and checked (e.g. non standard tensor shapes, etc).
+  Note that in most cases, such checks are already implemented (see above), but it's to the user to check that sufficient tests are checked.
+
+General technical details
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Implement the minimal change that fully satisfies the behavior described in a PR description or an issue;
+- Reuse existing abstractions in DeepInverse as much as possible to help modularity;
+- Anticipate potential future abstractions. For instance, if you propose a modification for some specific 1D application, try to open to door to 1D globally;
+- Avoid adding new dependencies unless the feature genuinely requires it;
```
Comment (nucli-vicky) [inline docs/source/contributing.rst]: maybe add
> and try to catch import errors with a 'to use ..., x is required. install it using ...'

File: docs/source/contributing.rst [hunk 5]
```diff
@@ -340,8 +424,20 @@ We get around this by mocking datasets in the tests. First, write the tests and
...
---------------------
+How to review PRs
+-----------------
+
+Reviewing PRs is a brilliant way to contribute to the DeepInverse community. Anyone can review PRs, especially if it covers your area of expertise. Here's a checklist for reviewers for all PRs:
+
+- Check that mathematical, methodological or algorithmic contributions are technically correct and match their relevant scientific publications;
+- Check that the author has followed the steps of the :ref:`contributing guide <step_by_step_contribute>`, including adding tests, appropriate docstrings, API and User Guide documentation, examples, and changelog;
+- Check that the contribution is allowed under the :ref:`LLM policy <llm-policy>`;
+- Check that the new code and documentation meets the :ref:`DeepInverse style guides <style_guides>` to ensure we maintain our high standard of code and docs.
```
Comment (nucli-vicky) [inline docs/source/contributing.rst:457]: additionally: check backward compatability + ways to deprecate instead of immediately breaking existing user-code.

**Review by nucli-vicky — CHANGES_REQUESTED** [review]
Thanks! Some thoughts / comments :)

File: docs/source/contributing.rst (hunk 1 above)
Comment (Tmodrzyk) [inline docs/source/contributing.rst]:
```suggestion
The DeepInverse community maintains a high, opinionated standard of code and documentation in order to provide a didactic library that leads the field of imaging, rather than a collection of code files.
```

File: docs/source/contributing.rst [hunk 6]
```diff
@@ -226,30 +230,56 @@ DeepInverse allows contributions where code is partially written by an LLM. Howe
...
 
-- Docstrings can be split into multiple sections using the horizontal separator ``|sep|``, with section titles introduced by ``:Title:``.
+- Split docstrings into multiple sections using the horizontal separator `|sep|`, and introduce section titles with `:Title:`.
+
+- To provide usage examples, include an `:Example:` section. Code in this section will be executed during documentation generation.
+
+- Use `:math:` for inline LaTeX-style mathematics, and `.. math::` for block equations.
 
-- To provide usage examples, include an ``:Example:`` section. Code in this section will be executed during documentation generation.
+- To include remarks, warnings, or tips, use the `.. note::` directive.
```
Comment (Tmodrzyk) [inline docs/source/contributing.rst]:
```suggestion
- To include remarks, warnings, or tips, use the `.. note::`, `.. warning::` or `.. tip::`  directives.
```

File: docs/source/contributing.rst (hunk 3 above)
Comment (Tmodrzyk) [reply docs/source/contributing.rst:328]: Indeed but that's still worth correcting for new examples.
I don't think that's the only guideline that is not followed across the full codebase.
We might want to do a pass for each of these in the current code and make sure we are compliant with what we ask. 😄

File: docs/source/contributing.rst [hunk 7]
```diff
@@ -294,8 +321,65 @@ Below is a minimal working example of a typical docstring that includes all thes
...
+- **Typing**: use concise modern typing as much as possible, but prefer readability over completeness. Where adding typing or more specific types would hurt readability, don't add it;
+- **Tests**: prefer adding a method / class to an existing generic test through a registry rather than a new standalone test;
+- **Tests**: the purpose of tests is that it checks the code does what it claims, not that the code executes without raising errors.
+  Therefore, adding an optimisation method requires to check that it converges to the limit point; adding a neural network from an external library should check metric on a dataset, etc. Limit cases should be added and checked (e.g. non standard tensor shapes, etc).
+  Note that in most cases, such checks are already implemented (see above), but it's to the user to check that sufficient tests are checked.
+
+General technical details
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Implement the minimal change that fully satisfies the behavior described in a PR description or an issue;
```
Comment (Tmodrzyk) [inline docs/source/contributing.rst]:
```suggestion
- Implement the minimal change that fully satisfies the behavior described in a Pull Request description or an issue;
```

File: docs/source/contributing.rst [hunk 8]
```diff
@@ -294,8 +321,65 @@ Below is a minimal working example of a typical docstring that includes all thes
...
+- **Cleanliness**: no global constants unless absolutely necessary; no unnecessary private single-use funcs; inline comments should only be used for describing non-obvious functionality, pitfalls or edge cases.
+- **Typing**: use concise modern typing as much as possible, but prefer readability over completeness. Where adding typing or more specific types would hurt readability, don't add it;
+- **Tests**: prefer adding a method / class to an existing generic test through a registry rather than a new standalone test;
+- **Tests**: the purpose of tests is that it checks the code does what it claims, not that the code executes without raising errors.
+  Therefore, adding an optimisation method requires to check that it converges to the limit point; adding a neural network from an external library should check metric on a dataset, etc. Limit cases should be added and checked (e.g. non standard tensor shapes, etc).
+  Note that in most cases, such checks are already implemented (see above), but it's to the user to check that sufficient tests are checked.
+
+General technical details
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Comment (Tmodrzyk) [inline docs/source/contributing.rst:376]: Here could we add something about PR length, especially for first-time contributors ? 
Not to set a hard limit, but I think everyone would probably agree that more than 3000 lines is too high. 
Since agents help so much, it's easy to get carried away and submit huge PRs.

File: docs/source/contributing.rst [hunk 9]
```diff
@@ -294,8 +321,65 @@ Below is a minimal working example of a typical docstring that includes all thes
...
+- Cite code whenever possible; see citation instructions :ref:`above <docstring_guidelines>`;
+- When copying code from external codebases, seek permission from the original author and include a `third-party licence <https://github.com/deepinv/deepinv/tree/main/deepinv/models/third_party>`_;
+- `nn.Modules` like physics should not possess device and dtype attributes; only tensors, parameters and buffers have devices and dtypes;
+- No mutable default arguments such as dicts;
+- Every operation should be batched.
+- If you propose a new technical convention, add it to this list so that future contributors and reviewers don't forget it!
+
 Backwards Compatibility
-------------------------
+~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (Tmodrzyk) [inline docs/source/contributing.rst:391]: I liked the item list style of the previous guidelines, can we have the same here ? I know this is from a previous PR but I don't think the content should be adapted, just the format.

File: docs/source/contributing.rst [hunk 10]
```diff
@@ -340,8 +424,20 @@ We get around this by mocking datasets in the tests. First, write the tests and
 Once the implementation, the tests and the mocking code are written, that they pass locally and on the CI servers, the maintainers will be able to review the code and merge it into the main branch if everything goes well. You should bear in mind though that the maintainers won't have the time to make sure the tests pass on the real data, so they will have to trust that you did things correctly.
 
 
-Maintainers commands
---------------------
+How to review PRs
+-----------------
+
+Reviewing PRs is a brilliant way to contribute to the DeepInverse community. Anyone can review PRs, especially if it covers your area of expertise. Here's a checklist for reviewers for all PRs:
+
```
Comment (Tmodrzyk) [inline docs/source/contributing.rst:452]: Could we add a small point on kindness / respect / well-meaning for PR comments ? 
Maybe a link to https://google.github.io/eng-practices/review/reviewer/comments.html for more precise indications?

**Review by Tmodrzyk — COMMENTED** [review]
Awesome @Andrewwango thank you for putting this together in the docs ! A few comments / additions

File: docs/source/contributing.rst (hunk 9 above)
Comment (Andrewwango) [reply docs/source/contributing.rst]: @jscanvic ?

File: docs/source/contributing.rst (hunk 10 above)
Comment (Andrewwango) [reply docs/source/contributing.rst:452]: thanks

File: docs/source/contributing.rst (hunk 2 above)
Comment (Andrewwango) [reply docs/source/contributing.rst:267]: thanks

File: docs/source/contributing.rst (hunk 5 above)
Comment (Andrewwango) [reply docs/source/contributing.rst:457]: thanks

File: docs/source/contributing.rst (hunk 9 above)
Comment (Andrewwango) [reply docs/source/contributing.rst:391]: done

**Review by Tmodrzyk — APPROVED** [review]
Awesome thank you again @Andrewwango !

File: docs/source/contributing.rst [hunk 11]
```diff
@@ -226,30 +230,58 @@ DeepInverse allows contributions where code is partially written by an LLM. Howe
...
-- To provide usage examples, include an ``:Example:`` section. Code in this section will be executed during documentation generation.
+- Use `:math:` for inline LaTeX-style mathematics, and `.. math::` for block equations.
 
-- Use ``:math:`` for inline LaTeX-style mathematics, and ``.. math::`` for block equations.
+- To include remarks, warnings, or tips, use the `.. note::`, `.. warning::` or `.. tip::`  directives.
 
-- To include remarks, warnings, or tips, use the ``.. note::`` directive.
+- Link objects with Sphinx roles such as `:class:`, `:func:`, `:meth:`, and `:ref:`.
+
+- Use single tick marks ` for inline code.
```
Comment (romainvo) [inline docs/source/contributing.rst:284]: Sorry, random comment. Shouldn't it be double backquotes for text rendered as monospaced text, i.e. inline code, and single backquotes for interpreted text, e.g. links, Sphinx markups, etc:
- https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#inline-markup
- https://docutils.sourceforge.io/docs/user/rst/quickref.html#inline-markup)

Agreed it's not obvious as most people only do markdown and single/double backquotes are equivalent..

Comment (Andrewwango) [reply docs/source/contributing.rst:284]: hmm dunno, this is what @tachella said in #379, so I thought single backquotes was convention for both code and sphinx interpreted text:

> - Set single backquotes `` `code here` ``as default for code mode instead of italics, as many contributors prefer this way since it matches the markdown default. Double backquotes will keep working as well for code mode.
